src/core/decorators/resolveUnderlyingPositions.ts

```diff
@@ -78,23 +70,34 @@ async function recursivePositionSolver({
   tokenPositions: (TokenBalance & { tokens?: Underlying[] })[]
   blockNumber?: number
 }) {
+  const promises = []
   for (const tokenPosition of tokenPositions) {
-    if (!tokenPosition.tokens) {
-      continue
-    }
+    promises.push(getUnderlyingAndRecurse(tokenPosition, adapter, blockNumber))
+  }
 
-    await resolveUnderlying({
-      underlying: tokenPosition.tokens,
-      adapter,
-      blockNumber,
-    })
+  await Promise.all(promises)
+}
 
-    await recursivePositionSolver({
-      adapter,
-      tokenPositions: tokenPosition.tokens,
-      blockNumber,
-    })
+async function getUnderlyingAndRecurse(
+  tokenPosition: (TokenBalance & { tokens?: Underlying[] }) | MovementsByBlock,
```

Comment:
```suggestion
  tokenPosition: (TokenBalance & { tokens?: Underlying[] }),
```

With the previous change, it is not necessary to add `MovementsByBlock`